idaes/core/util/model_diagnostics.py

```diff
@@ -3653,6 +3653,72 @@ def check_parallel_jacobian(model, tolerance: float = 1e-4, direction: str = "ro
 
     jac, nlp = get_jacobian(model, scaled=False)
 
+    # Get vectors that we will check, and the Pyomo components
+    # they correspond to.
+    if direction == "row":
+        components = nlp.get_pyomo_constraints()
+        mat = jac.tocsr()
+
+    elif direction == "column":
+        components = nlp.get_pyomo_variables()
+        mat = jac.transpose().tocsr()
+
+    norms = [norm(mat[i, :], ord="fro") for i in range(len(components))]
+
+    # Take product of all rows/columns with all rows/columns by taking outer
+    # product of matrix with itself
+    outer = mat @ mat.transpose()
```

Comment:
The previous implementation was designed to avoid taking every possible dot product by first sorting vectors by their nonzero structure. I would have expected this to be slow due to unnecessary dot products, but in some testing, it doesn't seem slower.

Reply:
This is sparse matrix multiplication carried out by Scipy. The key here is that 1) Sorting through which products to take should be carried out in the sparse matrix multiplication routine, not manually and 
2) The looping required to do this sorting and multiplication is done in compiled C++ code, not Python code, and (should) be much faster than doing it in Python code.

Reply:
The scipy matrix product is taking every possible combination of non-zero dot products, while the previous implementation takes only dot products between vectors with the same sparsity structure. This is fewer dot products, although, in practice, probably not by a wide margin.

Reply:
I'm afraid I don't entirely follow. Scipy's matrix multiply, as near as I can tell, [does the same thing.](https://github.com/scipy/scipy/blob/c241196b6b67343f90817d1800fab78d5736fb46/scipy/sparse/sparsetools/csr.h#L515) It implements the [SMMP algorithm](https://link.springer.com/article/10.1007/BF02070824) which occurs in two steps: the first to determine the nonzero structure of the matrix, the second to actually compute the values (as near as I can tell). That's basically what you're doing, except 1) it isn't filtering out entries as being "too small to contribute" and 2) it's doing it in C++ instead of Python, which is much faster.

Reply:
For example, if two rows are `(1, 2, 3)` and `(1, 2, 0)`, they cannot possibly be parallel. The previous implementation will not compute this dot product, while the new implementation will.

Reply:
So you're ruling out rows being parallel based on one having an element for the third index (3) and the other having zero for the third index? That can fail if the first row was `(1, 2, 3e-8)` instead of `(1, 2, 3)`: `(1, 2, 3e-8)` is still effectively parallel to `(1, 2, 0)` even if they structurally differ.

Reply:
We do this after filtering out small entries.